### ISSUE TITLE
Vault agent template does not execute command for all templates

### ISSUE THREAD

**Issue by herrbpl** [hashicorp/vault#8697]
**Describe the bug**
When you have multiple templates in vault agent config, like: 

```hcl
...
"template" = {
  "source" = "/vault/config/general.json.tpl"
  "destination" = "/vault/secrets/general/general.json"
  "error_on_missing_key"=true    
  "command" = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general"
}

# certificates

"template" = {
  "source" = "/vault/config/certificates.json.tpl"
  "destination" = "/vault/secrets/certificates/certificates.json"
  "error_on_missing_key"=true
  "command" = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates"
}

# Service credentials

"template" = {
  "source" = "/vault/config/credentials.json.tpl"
  "destination" = "/vault/secrets/credentials/credentials.json"
  "error_on_missing_key"=true
  "command" = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/credentials/credentials.json /vault/secrets/credentials"
}
...
```
then when agent is run and has rendered all the templates, it does not execute all commands but exits instead after last render has been completed, executing only one or two first commands.

**To Reproduce**
1. Create vault agent config where multiple templates are rendered and each template has command specified as in example, I have kubernetes role auth but use whatever method you like:
**config-init.hcl**
```hcl
"auto_auth" = {
  "method" = {
    "mount_path"= "auth/kubernetes"
    "config"= {
      "role"= "service-role"
    }        
    "type" = "kubernetes"
  }

  "sink" = {
    "config" = {
      "path" = "/home/vault/.token"
    }

    "type" = "file"
  }
}

"exit_after_auth" = true
"pid_file" = "/home/vault/.pid"

# general config

"template" = {
  "source" = "/vault/config/general.json.tpl"
  "destination" = "/vault/secrets/general/general.json"
  "error_on_missing_key"=true    
  "command" = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general"
}

# certificates

"template" = {
  "source" = "/vault/config/certificates.json.tpl"
  "destination" = "/vault/secrets/certificates/certificates.json"
  "error_on_missing_key"=true
  "command" = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates"
}

# Service credentials

"template" = {
  "source" = "/vault/config/credentials.json.tpl"
  "destination" = "/vault/secrets/credentials/credentials.json"
  "error_on_missing_key"=true
  "command" = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/credentials/credentials.json /vault/secrets/credentials"
}
    
"vault" = {
  "address" = "https://your_vault_instance:port"      
}
```
2. Run 
```bash
vault agent -config ./config-init.hcl
```
3. Only some commands get executed, it seems to depend on how fast rendering engine stops. 
```text
/ # vault agent -config ./config-init.hcl 
==> Vault server started! Log data will stream in below:

==> Vault agent configuration:

                     Cgo: disabled
               Log Level: info
                 Version: Vault v1.3.4

2020-04-08T09:10:28.606Z [INFO]  sink.file: creating file sink
2020-04-08T09:10:28.606Z [INFO]  sink.file: file sink configured: path=/home/vault/.token2 mode=-rw-r-----
2020-04-08T09:10:28.607Z [INFO]  auth.handler: starting auth handler
2020-04-08T09:10:28.607Z [INFO]  auth.handler: authenticating
2020-04-08T09:10:28.607Z [INFO]  template.server: starting template server
2020/04/08 09:10:28.608191 [INFO] (runner) creating new runner (dry: false, once: false)
2020-04-08T09:10:28.608Z [INFO]  sink.server: starting sink server
2020/04/08 09:10:28.608835 [INFO] (runner) creating watcher
2020-04-08T09:10:28.820Z [INFO]  auth.handler: authentication successful, sending token to sinks
2020-04-08T09:10:28.820Z [INFO]  auth.handler: starting renewal process
2020-04-08T09:10:28.821Z [INFO]  sink.file: token written: path=/home/vault/.token2
2020-04-08T09:10:28.821Z [INFO]  sink.server: sink server stopped
2020-04-08T09:10:28.821Z [INFO]  sinks finished, exiting
2020-04-08T09:10:28.821Z [INFO]  template.server: template server received new token
2020/04/08 09:10:28.821278 [INFO] (runner) stopping
2020/04/08 09:10:28.821408 [INFO] (runner) creating new runner (dry: false, once: false)
2020/04/08 09:10:28.821663 [INFO] (runner) creating watcher
2020/04/08 09:10:28.821870 [INFO] (runner) starting
2020-04-08T09:10:28.837Z [INFO]  auth.handler: renewed auth token
2020/04/08 09:10:28.937937 [INFO] (runner) rendered "/vault/config/general.json.tpl" => "/vault/secrets/general/general.json"
2020/04/08 09:10:28.938809 [INFO] (runner) executing command "/bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general" from "/vault/config/general.json.tpl" => "/vault/secrets/general/general.json"
2020/04/08 09:10:28.938872 [INFO] (child) spawning: /bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general
2020/04/08 09:10:29.414758 [INFO] (runner) rendered "/vault/config/certificates.json.tpl" => "/vault/secrets/certificates/certificates.json"
2020/04/08 09:10:29.415439 [INFO] (runner) rendered "/vault/config/credentials.json.tpl" => "/vault/secrets/credentials/credentials.json"
2020/04/08 09:10:29.415507 [INFO] (runner) executing command "/bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates" from "/vault/config/certificates.json.tpl" => "/vault/secrets/certificates/certificates.json"
2020/04/08 09:10:29.415569 [INFO] (child) spawning: /bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates
2020/04/08 09:10:29.415677 [INFO] (runner) stopping
2020-04-08T09:10:29.415Z [INFO]  template.server: template server stopped
```
Only two of three commands are executed (for general and certificates paths in this example).

**Expected behavior**
There should be THREE execution of command, however, there is only two before agent stops and exits. 

**Environment:**
* Vault CLI Version (retrieve with `vault version`): Vault v1.3.4
* Server Operating System/Architecture: alpine based vault docker hub image running on Kubernetes 1.17.6

**Comment by herrbpl** [comment]
Looks like problem is with multiple passing over template. This is content of credentials template. 

/ # cat /vault/config/credentials.json.tpl 
```hcl
{
  {{ scratch.Set "comma" "" }}
  {{ range secrets "secret/development/metadata/test-helm-tbs/paymentgateway-service/credentials" }}
  {{ scratch.Get "comma" }}
  "{{ . }}": 
      {{ with secret (printf "secret/development/test-helm-tbs/paymentgateway-service/credentials/%s" .)  }}{{ toJSONPretty .Data.data }}{{ end }}{{ scratch.Set "comma" "," }}
  {{ end }}
} 
``` 



As a workaround, I added extra template for rendering:
```hcl
# Workaround to vault bug

  "template" = {
    "source" = "/vault/config/credentials.json.tpl"
    "destination" = "/tmp/workaround.json"
    "error_on_missing_key"=true
    "wait" = "10s"
    "command" = "/bin/sh -c 'sleep 25'"
  }
```
This seems to trigger command.

**Comment by herrbpl** [comment]
It didn't seem to solve the issue. Looks like some optimization is running command per source file and not per target file. If I have one source template file which is rendered to two different targets then only one of target gets rendered and only one command gets executed, not both, even if i delete one target file, it will not be re-rendered by agent. 

This seems to be different bug from originally reported one.

**Comment by tyrannosaurus-becks** [comment]
Hi! Thanks for posting this issue. 

I dug through our templating code to see if we supported multiple templates, and it looks like we do. We have some test coverage on it where we parse [this template](https://github.com/hashicorp/vault/blob/master/command/agent/config/test-fixtures/config-template-many.hcl), which includes multiple values, and then check to make sure we receive what we expect. To check through what you're seeing, to the second template shown I added a new `command = "buzz"` line, then added a check for it to [this test](https://github.com/hashicorp/vault/blob/master/command/agent/config/config_test.go#L464). The test succeeded, indicating that in some situations it's possible to get the command through.

Would you be willing to convert your config to HCL, and model off that example template more to see if it resolves the issue for you? Thank you!

**Comment by herrbpl** [comment]
*NB!*  Problem only happens with templates, which have such loop to get first secrets and then contents as in following template

`cat credentials.json.tpl`: 

```hcl
    {
      {{ "{{" }} scratch.Set "comma" "" {{ "}}" }}
      {{ "{{" }} range secrets "{{ $secretpath_metadata_service }}/credentials" {{ "}}" }}
      {{ "{{" }} scratch.Get "comma" {{ "}}" }}
      "{{ "{{ . }}" }}": 
          {{ "{{" }} with secret (printf "{{ $secretpath_service }}/credentials/%s" .) {{ " }}{{ toJSONPretty .Data.data }}{{ end }}" }}{{ "{{" }} scratch.Set "comma" "," {{ "}}" }}
      {{ "{{ end }}" }}
    }  
```


And it seems to happen only when vault runs only once (exit_after_auth = true) or on first iteration (when exit_after_auth = false) - like it runs command before it has really rendered template fully, it runs already on first pass! 

But since output does not contain output yet, command processes empty file and appears not to working.

 When vault has been running and has populated dependency tree, then command appears to work - but it really uses results of previous run.

Regular templates which retrieve only secret, seem to run without issues.
 Converted to HCL format as suggested, result is same. Configuration and logs follow:
`cat config.hcl`:
```hcl
auto_auth  {
  method  {
    mount_path = "auth/dev-k8s-02"
    config  {
      role = "test-helm-tbs-paymentgateway-service"
    }        
    type = "kubernetes"
  }

  sink {
    config  {
      "path" = "/home/vault/.token"
    }

    type = "file"
  }
}

exit_after_auth = false
pid_file = "/home/vault/.pid"

# general config

template  {
  source = "/vault/config/general.json.tpl"
  destination = "/vault/secrets/general/general.json"
  error_on_missing_key = true    
  command = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general"
}

# certificates

template {
  source = "/vault/config/certificates.json.tpl"
  destination = "/vault/secrets/certificates/certificates.json"
  error_on_missing_key = true
  command = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates true"
}

# Service credentials

template {
  source = "/vault/config/credentials.json.tpl"
  destination = "/vault/secrets/credentials/credentials.json"
  error_on_missing_key = true
  command = "/bin/sh /vault/config/splitconfig.sh /vault/secrets/credentials/credentials.json /vault/secrets/credentials"
}

vault = {
  address = "https://my.vault.address:port"      
}
```
log:
```text
==> Vault agent configuration:
Cgo: disabled
Log Level: info
Version: Vault v1.3.4
2020-04-09T07:20:20.030Z [INFO] sink.file: creating file sink
2020-04-09T07:20:20.031Z [INFO] sink.file: file sink configured: path=/home/vault/.token mode=-rw-r-----
2020-04-09T07:20:20.031Z [INFO] auth.handler: starting auth handler
2020-04-09T07:20:20.031Z [INFO] auth.handler: authenticating
2020-04-09T07:20:20.032Z [INFO] sink.server: starting sink server
2020-04-09T07:20:20.032Z [INFO] template.server: starting template server
2020/04/09 07:20:20.032883 [INFO] (runner) creating new runner (dry: false, once: false)
2020/04/09 07:20:20.033394 [INFO] (runner) creating watcher
2020-04-09T07:20:20.215Z [INFO] auth.handler: authentication successful, sending token to sinks
2020-04-09T07:20:20.215Z [INFO] auth.handler: starting renewal process
2020-04-09T07:20:20.216Z [INFO] sink.file: token written: path=/home/vault/.token
2020-04-09T07:20:20.216Z [INFO] sink.server: sink server stopped
2020-04-09T07:20:20.216Z [INFO] sinks finished, exiting
2020-04-09T07:20:20.216Z [INFO] template.server: template server received new token
2020/04/09 07:20:20.216039 [INFO] (runner) stopping
2020/04/09 07:20:20.216069 [INFO] (runner) creating new runner (dry: false, once: false)
2020/04/09 07:20:20.216214 [INFO] (runner) creating watcher
2020/04/09 07:20:20.216336 [INFO] (runner) starting
2020-04-09T07:20:20.228Z [INFO] auth.handler: renewed auth token
2020/04/09 07:20:20.333464 [INFO] (runner) rendered "/vault/config/general.json.tpl" => "/vault/secrets/general/general.json"
2020/04/09 07:20:20.333975 [INFO] (runner) executing command "/bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general" from "/vault/config/general.json.tpl" => "/vault/secrets/general/general.json"
2020/04/09 07:20:20.334033 [INFO] (child) spawning: /bin/sh /vault/config/splitconfig.sh /vault/secrets/general/general.json /vault/secrets/general
2020/04/09 07:20:21.058773 [INFO] (runner) rendered "/vault/config/certificates.json.tpl" => "/vault/secrets/certificates/certificates.json"
2020/04/09 07:20:21.059024 [INFO] (runner) executing command "/bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates true" from "/vault/config/certificates.json.tpl" => "/vault/secrets/certificates/certificates.json"
2020/04/09 07:20:21.059077 [INFO] (child) spawning: /bin/sh /vault/config/splitconfig.sh /vault/secrets/certificates/certificates.json /vault/secrets/certificates true
2020/04/09 07:20:21.453857 [INFO] (runner) rendered "/vault/config/credentials.json.tpl" => "/vault/secrets/credentials/credentials.json"
2020/04/09 07:20:21.453903 [INFO] (runner) executing command "/bin/sh /vault/config/splitconfig.sh /vault/secrets/credentials/credentials.json /vault/secrets/credentials" from "/vault/config/credentials.json.tpl" => "/vault/secrets/credentials/credentials.json"
2020/04/09 07:20:21.453954 [INFO] (child) spawning: /bin/sh /vault/config/splitconfig.sh /vault/secrets/credentials/credentials.json /vault/secrets/credentials
2020/04/09 07:20:21.454148 [INFO] (runner) stopping
2020-04-09T07:20:21.454Z [INFO] template.server: template server stopped
```
It runs now three instances of command. But end result is still wrong as command is run before template with loop is fully rendered.

**Comment by herrbpl** [comment]
I added copying of rendered outputfile contents tobackup location.

From three templates, 100% copied were only teplates without "secrets" loop in them, other copies of files showed up inconsitantly which  indicates that command is not actually executing.

**Comment by tyrannosaurus-becks** [comment]
:+1: gotcha. Thank you so much for all these steps to reproduce, very helpful!

**Comment by heatherezell** [comment]
Thanks for the report! This issue was filed against Vault `v1.3.4`, which is no longer a supported release.

We're unable to investigate bugs against unsupported versions. If you're still experiencing this on a current supported release of Vault, please open a new issue with updated environment details and we'll take a look.

For a list of currently supported versions, see the [Vault version support page](https://github.com/hashicorp/vault/blob/main/.release/versions.hcl). For Enterprise long-term support information, see the [LTS tracker](https://developer.hashicorp.com/vault/docs/updates/lts-tracker).
